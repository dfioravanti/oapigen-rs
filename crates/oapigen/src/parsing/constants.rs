use oas3::spec::ObjectSchema;
use serde_json::Value;

use crate::models;

fn parse_constants(schema_name: String, schema: ObjectSchema) -> models::TokenizedSchema {
    match schema.const_value.unwrap() {
        Value::Null => todo!(),
        Value::Bool(v) => todo!(),
        Value::Number(v) => todo!(),
        Value::String(v) => todo!(),
        Value::Array(values) => todo!(),
        Value::Object(field_to_value) => todo!(),
    }
}

fn parse_value(original_value: Value) {
    let mut stack = Vec::new();

    stack.push(original_value);

    while let Some(value) = stack.pop() {}
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    #[rstest]
    #[case(
        "string",
        "{{\"country\": {{\"const\": \"United States of America\"}}}}\""
    )]
    fn test_parse_base_cases() {}
}
