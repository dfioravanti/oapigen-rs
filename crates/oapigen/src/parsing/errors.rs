use oas3::spec;
use proc_macro2::LexError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParsingError {
    #[error("could not parse the string to TokenStream.")]
    CannotParseString(#[from] LexError),
    #[error("could not resolve the reference with the passed spec.")]
    ReferenceNotFound(#[from] spec::RefError),
}
