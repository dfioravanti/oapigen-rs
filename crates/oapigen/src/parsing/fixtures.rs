use indoc::indoc;
use rstest::fixture;

#[fixture]
fn twenty_one() -> &'static str {
    indoc! {r#"
        def hello():
            print("Hello, world!")

        hello()
    "#}
}
