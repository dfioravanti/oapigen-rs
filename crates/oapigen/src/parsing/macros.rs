use crate::models;
use proc_macro2::TokenStream;
use quote::quote;
use std::collections::HashMap;

const NAME_IMPORTS_SERDE: &str = "SERDE";

fn get_default_macros() -> (TokenStream, models::Imports) {
    let mut imports = HashMap::new();
    imports.insert(
        NAME_IMPORTS_SERDE.to_string(),
        quote! { use serde::{Deserialize, Serialize}; },
    );

    let default_macros = quote! {
        #[derive(Debug, Deserialize, Serialize)]
    };

    (default_macros, imports)
}

/// Returns the macros that are associated with the current type.
/// By default, we add derive(Debug, Deserialize, Serialize) to allow for reasonable behaviour.
pub(crate) fn get_macros() -> (TokenStream, models::Imports) {
    get_default_macros()
}
