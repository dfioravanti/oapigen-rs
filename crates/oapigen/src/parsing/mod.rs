mod constants;
pub mod errors;
mod fixtures;
mod macros;
mod routes;
mod schemas;
pub mod specs;
mod typeset;
