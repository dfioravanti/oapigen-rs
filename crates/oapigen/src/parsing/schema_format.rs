//! [https://spec.openapis.org/registry/format/]

use std::collections::HashMap;

use crate::models;
use log::warn;
use proc_macro2::TokenStream;
use quote::quote;

pub fn default_string() -> (TokenStream, models::Imports) {
    (quote! { String }, HashMap::new())
}

/// Formats strings according to the registry provided in
/// [https://spec.openapis.org/registry/format/].
///
/// The list of supported formats is a subset of the registry.
pub fn formatted_string(type_format: &str) -> (TokenStream, models::Imports) {
    let tokenized_type: TokenStream;
    let mut imports = HashMap::new();

    match type_format {
        "time" => {
            tokenized_type = quote! { DateTime };
            imports.insert(
                "chrono::DateTime".to_string(),
                quote! { use chrono::DateTime },
            );
        }
        _ => {
            warn!("format {type_format} is unknown for strings, defaulting to string");
            return default_string();
        }
    }

    (tokenized_type, imports)
}

pub fn default_number() -> (TokenStream, models::Imports) {
    (quote! { f32 }, HashMap::new())
}

/// Formats numbers (including integers) according to the registry provided in
/// [https://spec.openapis.org/registry/format/].
///
/// The list of supported formats is a subset of the registry.
pub(crate) fn formatted_number(type_format: &str) -> (TokenStream, models::Imports) {
    let tokenized_type: TokenStream;
    let imports = HashMap::new();

    match type_format {
        "int64" => tokenized_type = quote! { i64 },
        "integer" | "int32" => tokenized_type = quote! { i32 },
        "int16" => tokenized_type = quote! { i16 },
        "int8" => tokenized_type = quote! { i8 },
        "uint64" => tokenized_type = quote! { u64 },
        "uint32" => tokenized_type = quote! { u32 },
        "uint16" => tokenized_type = quote! { u16 },
        "uint8" => tokenized_type = quote! { u8 },
        "double" => tokenized_type = quote! { f64 },
        _ => {
            warn!("format {type_format} is unknown for numbers, defaulting to number");
            return default_number();
        }
    }

    (tokenized_type, imports)
}
