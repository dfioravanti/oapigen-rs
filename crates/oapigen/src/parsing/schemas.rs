use crate::parsing::errors;
use oas3::spec;
use oas3::spec::{SchemaType, SchemaTypeSet};

pub fn parse_schema(schema: spec::ObjectSchema) {
    schema.schema_type;
}

fn validate_schema(schema: spec::ObjectSchema) -> Result<(), errors::ParsingError> {
    Ok(())
}

fn is_type(typeset: SchemaTypeSet, schema_type: SchemaType) -> bool {
    match typeset {
        SchemaTypeSet::Single(t) => t.eq(&schema_type),
        SchemaTypeSet::Multiple(items) => items.iter().any(|t| t.eq(&schema_type)),
    }
}
