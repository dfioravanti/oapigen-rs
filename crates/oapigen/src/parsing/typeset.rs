use crate::{models, parsing::schema_format};
use oas3::spec::{ObjectSchema, SchemaType, SchemaTypeSet};
use proc_macro2::TokenStream;
use quote::ToTokens;

pub fn tokenize_schema(schema_name: String, schema: ObjectSchema) -> TokenStream {
    if schema.const_value.is_some() {}

    let tokenized_schema = match &schema.schema_type {
        Some(schema_typeset) => match schema_typeset {
            SchemaTypeSet::Single(schema_type) => {
                tokenize_flat_schema(schema_name, &schema, schema_type)
            }
            SchemaTypeSet::Multiple(_items) => todo!(),
        },
        None => todo!(),
    };

    tokenized_schema.to_token_stream()
}

fn tokenize_flat_schema(
    schema_name: String,
    schema: &ObjectSchema,
    schema_type: &SchemaType,
) -> models::TokenizedSchema {
    match schema_type {
        SchemaType::Boolean => todo!(),
        SchemaType::Integer => tokenize_integer_schema(schema_name, schema),
        SchemaType::Number => tokenize_number_schema(schema_name, schema),
        SchemaType::String => tokenize_string_schema(schema_name, schema),
        SchemaType::Array => todo!(),
        SchemaType::Object => todo!(),
        SchemaType::Null => todo!(),
    }
}

fn tokenize_integer_schema(schema_name: String, schema: &ObjectSchema) -> models::TokenizedSchema {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse().unwrap();

    let (tokenized_type, imports) = match &schema.format {
        Some(format) => schema_format::formatted_number(format),
        None => schema_format::formatted_number("integer"),
    };

    models::TokenizedSchema {
        tokenized_name,
        tokenized_type,
        imports,
        current_type: models::CurrentType::Type,
    }
}

fn tokenize_number_schema(schema_name: String, schema: &ObjectSchema) -> models::TokenizedSchema {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse().unwrap();

    let (tokenized_type, imports) = match &schema.format {
        Some(format) => schema_format::formatted_number(format),
        None => schema_format::default_number(),
    };

    models::TokenizedSchema {
        tokenized_name,
        tokenized_type,
        imports,
        current_type: models::CurrentType::Type,
    }
}

fn tokenize_string_schema(schema_name: String, schema: &ObjectSchema) -> models::TokenizedSchema {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse().unwrap();

    let (tokenized_type, imports) = match &schema.format {
        Some(format) => schema_format::formatted_string(format),
        None => schema_format::default_string(),
    };

    models::TokenizedSchema {
        tokenized_name,
        tokenized_type,
        imports,
        current_type: models::CurrentType::Type,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::parsing::fixtures;
    use indoc::indoc;
    use rstest::rstest;

    #[rstest]
    #[case("integer", "Age", "type: integer", "type Age = i32 ;")]
    #[case("number", "Height", "type: number", "type Height = f32 ;")]
    #[case("string", "Name", "type: string", "type Name = String ;")]
    fn test_parse_base_cases(
        #[case] name: &str,
        #[case] schema_name: &str,
        #[case] schema_spec: &str,
        #[case] expected: &str,
    ) {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(name);

        let schema = serde_yaml::from_str::<ObjectSchema>(schema_spec).unwrap();

        let got = tokenize_schema(schema_name.to_string(), schema);

        settings.bind(|| {
            insta::assert_yaml_snapshot!(got.to_string());
        });
        assert_eq!(got.to_string(), expected.to_string());
    }

    #[rstest]
    #[case("integer", "Age", "const: 3", "type Age = i32 ;")]
    fn test_parse_constants(
        #[case] name: &str,
        #[case] schema_name: &str,
        #[case] schema_spec: &str,
        #[case] expected: &str,
    ) {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(name);

        let schema = serde_yaml::from_str::<ObjectSchema>(schema_spec).unwrap();

        let got = tokenize_schema(schema_name.to_string(), schema);

        settings.bind(|| {
            insta::assert_yaml_snapshot!(got.to_string());
        });
        assert_eq!(got.to_string(), expected.to_string());
    }
}
