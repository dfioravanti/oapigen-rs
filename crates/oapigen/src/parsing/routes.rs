use crate::models;
use crate::models::schema;
use crate::parsing::errors::ParsingError;
use crate::parsing::typeset::tokenize_schema;
use convert_case::{Case, Casing};
use oas3::spec;
use std::option::Option;

pub(crate) fn parse_routes(
    config: &models::Config,
    spec: &oas3::Spec,
) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut output = Vec::new();

    let paths = &spec.paths;

    if let Some(paths) = paths {
        for (route_name, path) in paths {
            if path.get.is_some() {
                let parsed =
                    parse_operation(config, spec, &"Get".to_string(), route_name, &path.get)?;
                output.extend(parsed);
            }
        }
    }

    Ok(output)
}

fn parse_operation(
    config: &models::Config,
    spec: &oas3::Spec,
    method_name: &String,
    route_name: &String,
    operation: &Option<spec::Operation>,
) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut output = Vec::new();
    if let Some(operation) = operation {
        let operation_name = match &operation.operation_id {
            Some(operation_id) => operation_id.to_case(Case::UpperCamel),
            None => [method_name, &route_name.to_case(Case::UpperCamel)]
                .map(String::as_str)
                .concat(),
        };

        let responses = &operation.responses;
        if let Some(responses) = responses {
            for (response_name, response) in responses {
                let resolved_response = response.resolve(spec)?;
                let parsed = parse_response(
                    config,
                    spec,
                    &operation_name,
                    &response_name,
                    &resolved_response,
                )?;
                output.extend(parsed);
            }
        }
    }
    Ok(output)
}

fn parse_response(
    config: &models::Config,
    spec: &oas3::Spec,
    operation_name: &String,
    response_name: &String,
    response: &spec::Response,
) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut v = Vec::with_capacity(response.content.len());
    for (mediatype_name, mediatype) in &response.content {
        let parsed = parse_mediatype(config, spec, operation_name, response_name, mediatype)?;
        match parsed {
            Some(tokenized_schema) => v.push(tokenized_schema),
            None => {}
        }
    }

    Ok(v)
}

fn parse_mediatype(
    config: &models::Config,
    spec: &oas3::Spec,
    operation_name: &String,
    response_name: &String,
    media_type: &spec::MediaType,
) -> Result<Option<schema::SchemaAsRust>, ParsingError> {
    let schema_name = [
        operation_name.to_case(Case::UpperCamel),
        "Response".to_string(),
        response_name.to_case(Case::UpperCamel),
    ]
    .join("");

    let schema = &media_type.schema;
    if let Some(schema) = schema {
        let parsed_schema = schema.resolve(spec)?;
        let tokens = tokenize_schema(config, schema_name, parsed_schema)?;

        return Ok(Some(tokens));
    }

    Ok(None)
}
