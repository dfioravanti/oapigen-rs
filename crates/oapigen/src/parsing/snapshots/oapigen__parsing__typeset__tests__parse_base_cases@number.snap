---
source: crates/oapigen/src/parsing/typeset.rs
expression: got.to_string()
---
use serde::{Deserialize, Serialize};
#[derive(Debug, Deserialize, Serialize)]
struct Height(f32);
